mlir/lib/Conversion/ConvertToAIRPass.cpp

```diff
@@ -1539,88 +1171,6 @@ struct CopyToDmaPass : public air::impl::CopyToDmaBase<CopyToDmaPass> {
-        dst_condense |= isa<memref::SubViewOp>(dst_defop);
-      }
-      if (src_condense || dst_condense) {
-        // Fields in the tuple: (1) dma op, (2) list of memref ops producing the
-        // src memref, and (3) list of memref ops producing the dst memref.
-        std::tuple<air::DmaMemcpyNdOp, std::vector<Operation *>,
-                   std::vector<Operation *>>
-            log_entry;
-        std::get<0>(log_entry) = dmaOp;
-        if (src_condense) {
-          Operation *ancestor = dmaOp.getSrcMemref().getDefiningOp();
-          bool exit = false;
-          while (ancestor && !exit) {
-            if (auto transpose_anc = dyn_cast<memref::TransposeOp>(ancestor)) {
-              std::get<1>(log_entry).push_back(ancestor);
-              ancestor = transpose_anc.getIn().getDefiningOp();
-            } else if (auto expand_anc =
-                           dyn_cast<memref::ExpandShapeOp>(ancestor)) {
-              std::get<1>(log_entry).push_back(ancestor);
-              ancestor = expand_anc.getSrc().getDefiningOp();
-            } else if (auto subview_anc =
-                           dyn_cast<memref::SubViewOp>(ancestor)) {
-              std::get<1>(log_entry).push_back(ancestor);
-              ancestor = subview_anc.getSource().getDefiningOp();
-            } else if (auto cast_anc = dyn_cast<memref::CastOp>(ancestor)) {
-              std::get<1>(log_entry).push_back(ancestor);
-              ancestor = cast_anc.getViewSource().getDefiningOp();
-            } else
-              exit = true;
-          }
-        }
-        if (dst_condense) {
-          Operation *ancestor = dmaOp.getDstMemref().getDefiningOp();
-          bool exit = false;
-          while (ancestor && !exit) {
-            if (auto transpose_anc = dyn_cast<memref::TransposeOp>(ancestor)) {
-              std::get<2>(log_entry).push_back(ancestor);
-              ancestor = transpose_anc.getIn().getDefiningOp();
-            } else if (auto expand_anc =
-                           dyn_cast<memref::ExpandShapeOp>(ancestor)) {
-              std::get<2>(log_entry).push_back(ancestor);
-              ancestor = expand_anc.getSrc().getDefiningOp();
-            } else if (auto subview_anc =
-                           dyn_cast<memref::SubViewOp>(ancestor)) {
-              std::get<2>(log_entry).push_back(ancestor);
-              ancestor = subview_anc.getSource().getDefiningOp();
-            } else if (auto cast_anc = dyn_cast<memref::CastOp>(ancestor)) {
-              std::get<2>(log_entry).push_back(ancestor);
-              ancestor = cast_anc.getViewSource().getDefiningOp();
-            } else
-              exit = true;
-          }
-        }
-        dma_ops.push_back(log_entry);
-      }
-    });
-    for (auto dmaOp : dma_ops) {
-      if (failed(condenseMemrefDataReorderingToAIRDma(
-              std::get<0>(dmaOp), std::get<1>(dmaOp), std::get<2>(dmaOp)))) {
-        return signalPassFailure();
-      }
-    }
```

Comment:
Run the moved code here?
```suggestion
    RewritePatternSet pattern(ctx);
    DmaMemcpyNdOp::getCanonicalizationPatterns(pattern, ctx);
    (void)applyPatternsAndFoldGreedily(module, std::move(pattern));
```

Reply:
Great point. Updated.